amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ConsumerModeUpdateIntegrationTest.java

```diff
@@ -66,218 +85,352 @@ public static Collection<Object[]> data() {
+                    nextMode);
+            try {
+                helper.runProducerAndConsumersThenWait(Lists.newArrayList(helper.consumerA, helper.consumerBPrev));
+                helper.stopProducerAndConsumers(Lists.newArrayList(helper.consumerBPrev));
+                assertTrue(helper.validateRecordsDelivery());
+
+                helper.runProducerAndConsumersThenWait(Lists.newArrayList(helper.consumerA, helper.consumerBNext));
+                helper.stopProducerAndConsumers(Lists.newArrayList(helper.consumerBNext, helper.consumerA));
+                assertTrue(helper.validateRecordsDelivery());
+                verifyLeasesType(helper.prevKCLAppConfig.buildAsyncDynamoDbClient(),
+                        KCLAppConfig.INTEGRATION_TEST_RESOURCE_PREFIX + helper.testName,
+                        helper.consumerA.getWorkerIdentifier(), prevMode, helper.consumerBNext.getWorkerIdentifier(),
+                        nextMode);
+
+                helper.runProducerAndConsumersThenWait(Lists.newArrayList(helper.consumerA));
+                helper.stopProducerAndConsumers(Lists.newArrayList(helper.consumerA));
+                assertFalse(helper.validateRecordsDelivery());
+                assertTrue(getLeaseCountOwnedBy(helper.prevKCLAppConfig.buildAsyncDynamoDbClient(),
+                        KCLAppConfig.INTEGRATION_TEST_RESOURCE_PREFIX + helper.testName,
+                        helper.consumerBNext.getWorkerIdentifier()) > 0);
+            } finally {
+                // Clean up
+                helper.cleanup();
             }
         }
+    }
 
-        private Boolean verifyLeaseType(Map<String, AttributeValue> lease, KCLMode mode) throws Exception {
-            AttributeValue streamName = lease.get("streamName");
-            AttributeValue shardId = lease.get("shardId");
-            switch (mode) {
-                case SingleStream:
-                    return streamName == null && shardId == null;
-                case SingleStreamCompatible:
-                    return true;
-                case SingleStreamUpgrade:
-                case MultiStream:
-                    return streamName != null && shardId != null;
-                default:
-                    throw new Exception("Unable to verify lease type with the given mode: " + mode);
+    private static void verifyLeasesType(DynamoDbAsyncClient ddbClient, String tableName, String consumerAIdentifier,
+            KCLMode consumerAMode, String consumerBIdentifier, KCLMode consumerBMode)
+            throws Exception {
+        ScanRequest scanRequest = ScanRequest.builder().tableName(tableName).build();
+        ScanResponse scanResponse = ddbClient.scan(scanRequest).get(30, TimeUnit.SECONDS);
+        for (Map<String, AttributeValue> item : scanResponse.items()) {
+            String leaseOwner = item.get("leaseOwner").s();
+            assertEquals(anyOf(is(consumerAIdentifier), is(consumerBIdentifier)), leaseOwner);
```

Comment:
My concern is that this test will also pass if workerA had all leases and workerB had none. Maybe we should also assert the minimum number of leases that should be assigned to each worker, or some other validation that's equally effective.

Reply:
`verifyLeaseType` only verify if leases are formatted as expected. Separate validation could be done using the function `getLeaseCountOwnedBy` for minimum number of leases owned by specific worker.